File: JS/CombinationSum.js
class CombinationSum {
    combinationSum(candidates, target) {
        const results = [];
        const comb = [];

        this.backtrack(target, comb, 0, candidates, results);
        return results;
    }

    backtrack(remain, comb, start, candidates, results) {
        if (remain === 0) {
            results.push([...comb]);
            return;
        } else if (remain < 0) {
            return;
        }

        for (let i = start; i < candidates.length; i++) {
            comb.push(candidates[i]);
            this.backtrack(remain - candidates[i], comb, i, candidates, results);
            comb.pop();
        }
    }
}

// Example usage:
const solution = new CombinationSum();
const candidates = [2, 3, 6, 7];
const target = 7;
const results = solution.combinationSum(candidates, target);

results.forEach(result => {
    console.log(result.join(' '));
});

File: Python/CombinationSum.py
class CombinationSum:
    def combinationSum(self, candidates, target):
        results = []
        comb = []

        def backtrack(remain, comb, start):
            if remain == 0:
                results.append(list(comb))
                return
            elif remain < 0:
                return

            for i in range(start, len(candidates)):
                comb.append(candidates[i])
                backtrack(remain - candidates[i], comb, i)
                comb.pop()

        backtrack(target, comb, 0)
        return results

# Example usage:
solution = CombinationSum()
candidates = [2, 3, 6, 7]
target = 7
results = solution.combinationSum(candidates, target)

for result in results:
    print(result)
